File: Data Structures & Algorithms/max-water-container/submission-1.java
class Solution {
    public int maxArea(int[] heights) {
        int left = 0, right = heights.length - 1, result = 0;
        while (left < right) {
            if (heights[left] > heights[right]) {
                result = Math.max(result, (right - left) * heights[right]);
                right--;
            } else {
                result = Math.max(result, (right - left) * heights[left]);
                left++;
            }
        } 
        return result;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        result = 0
        for i in range(len(heights)):
            start = i
            while stack and heights[i] < stack[-1][1]:
                temp = stack.pop()
                result = max(result, (i - temp[0]) * temp[1])
                start = temp[0]
            stack.append((start, heights[i]))
        while stack:
            temp = stack.pop()
            result = max(result, (len(heights) - temp[0]) * temp[1])
        return result